lib/Cron/Workers/WooCommerceSync.php

```diff
@@ -27,13 +29,22 @@ function checkProcessingRequirements() {
     return $this->woocommerce_helper->isWooCommerceActive();
   }
 
+  function prepareTask(ScheduledTask $task) {
+    if (is_null($task->status)) {
+      $this->taskAlreadyInProgress = true;
+      return false;
+    }
+
+    return parent::prepareTask($task);
+  }
+
   function processTaskStrategy(ScheduledTask $task) {
+    if ($this->taskAlreadyInProgress) {
```

Comment:
What will happen if the task fails (e.g. some mysql timeout or just some server memory issue)? I'm afraid that this task could be stuck in running state. Do we have some mechanism to mark task as timeouted and reschedule it?
I was not able to find something like that in SimpleWorker. Maybe we could add some execution timeout e.g. one day and in case the task is running for a day (is null and not updated for a day) then run it again and update tasks updated_at

Reply:
We don't have any mechanism for dealing with faulty tasks. I think a global mechanism is out of the scope of this issue (we need to consider its impact on all other tasks), but I've added a timeout check inside the WC sync task.

P.S.: I also found that the previous solution for ensuring a single instance didn't work because the flag was a class property, but `prepareTask` and `processTask` were called in different requests so the flag was never passed from one to another. Now I've added a `meta` field to scheduled tasks (like in sending queues) for persistent storage of this kind of data.